src/main/java/hudson/plugins/deploy/DeployPublisher.java

```diff
@@ -47,17 +46,43 @@ public DeployPublisher(List<ContainerAdapter> adapters, String war, String conte
         this.contextPath = contextPath;
     }
 
+    @SuppressFBWarnings("NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE")
     @Override
+    @Deprecated
     public boolean perform(AbstractBuild<?,?> build, Launcher launcher, BuildListener listener) throws InterruptedException, IOException {
-        if (build.getResult().equals(Result.SUCCESS) || onFailure) {
-            for (FilePath warFile : build.getWorkspace().list(this.war)) {
-                for (ContainerAdapter adapter : adapters)
-                    if (!adapter.redeploy(warFile, contextPath, build, launcher, listener))
-                        build.setResult(Result.FAILURE);
+        perform(false, build, build.getWorkspace(), launcher, listener);
+        return true;
+    }
+
+    @Override
+    public void perform(@Nonnull Run<?, ?> run, @Nonnull FilePath workspace, @Nonnull Launcher launcher, @Nonnull TaskListener listener) throws InterruptedException, IOException {
+        perform(true, run, workspace, launcher, listener);
+    }
+
+    private void perform (boolean fromWorkFlow, @Nonnull Run<?, ?> run, @Nonnull FilePath workspace, @Nonnull Launcher launcher, @Nonnull TaskListener listener) throws InterruptedException, IOException {
+        if (fromWorkFlow || onFailure || Result.SUCCESS.equals(run.getResult())) {
+            if (!workspace.exists()) {
+                listener.getLogger().println("[DeployPublisher][ERROR] Workspace not found");
+                throw new FileNotFoundException("Workspace not found");
             }
-        }
 
-        return true;
+            FilePath[] wars = workspace.list(this.war);
```

Comment:
Before https://github.com/jenkinsci/jenkins/pull/2914 in Jenkins 2.66 this method may actually return null. I would recommend explicitly  checking it